lib/src/back_end/solution.dart

```diff
@@ -130,43 +82,141 @@ class Solution implements Comparable<Solution> {
+  void invalidate(Piece piece) {
+    _invalidPiece = piece;
+  }
 
   /// When called on a [Solution] with some unselected piece states, chooses a
   /// piece and yields further solutions for each state that piece can have.
   List<Solution> expand(Piece root, int pageWidth) {
-    if (_nextPieceToExpand case var piece?) {
-      return [
-        for (var state in piece.states)
-          if (_stateSet.tryBind(piece, state) case var stateSet?)
-            Solution._(root, pageWidth, stateSet)
-      ];
+    // If the piece whose newline constraint was violated is already bound to
+    // one state, then every solution derived from this one will also fail in
+    // the same way, so discard the whole solution tree hanging off this one.
+    if (_invalidPiece case var piece? when isBound(piece)) return const [];
+
+    var expandPiece = _nextPieceToExpand;
+
+    // If there is no piece that we can expand on this solution, it's a dead
+    // end (or a winner).
+    if (expandPiece == null) return const [];
+
+    // TODO(perf): If `_invalidPiece == expandPiece`, then we know that the
+    // first state leads to an invalid solution, so there's no point in trying
+    // to expand to a solution that binds `expandPiece` to
+    // `expandPiece.states[0]`. We should be able to do:
+    //
+    //     Iterable<State> states = expandPiece.states;
+    //     if (_invalidPiece == expandPiece) {
+    //       print('skip $expandPiece ${states.first}');
+    //       states = states.skip(1);
+    //     }
+    //
+    // And then use `states` below. But when I tried that, it didn't seem to
+    // make any noticeable performance difference on the one pathological
+    // example I tried. Leaving this here as a TODO to investigate more when
+    // there are other benchmarks we can try.
+
+    return [
+      for (var state in expandPiece.states)
+        if (_tryBind(root, pageWidth, expandPiece, state) case var solution?)
+          solution
+    ];
+  }
+
+  /// Attempts to extend this solution's piece states by binding [piece] to
+  /// [state], taking into account any constraints pieces place on each other.
+  ///
+  /// Returns a new [Solution] with [piece] bound to [state] and any other
+  /// pieces constrained by that choice bound to their constrained values
+  /// (recursively). Returns `null` if a constraint conflicts with the already
+  /// bound or pinned state for some piece.
+  Solution? _tryBind(Piece root, int pageWidth, Piece piece, State state) {
+    var conflict = false;
+    var newStates = {..._pieceStates};
+    var newCost = cost;
+
+    void bind(Piece thisPiece, State thisState) {
+      // If this piece is already pinned or bound to some other state, then the
+      // solution doesn't make sense.
+      var alreadyBound = thisPiece.pinnedState ?? newStates[thisPiece];
+      if (alreadyBound != null && alreadyBound != thisState) {
+        conflict = true;
+        return;
+      }
```

Comment:
Would it make sense to check for conflicts in sibling branches and stop early?

```suggestion
      if (conflict) return;
      // If this piece is already pinned or bound to some other state, then the
      // solution doesn't make sense.
      var alreadyBound = thisPiece.pinnedState ?? newStates[thisPiece];
      if (alreadyBound != null && alreadyBound != thisState) {
        conflict = true;
        return;
      }
```

Reply:
I don't think it will make a measurable difference, but it doesn't hurt to early out. I'll make the change locally and add a comment.